lib/pdf/reader.rb

```diff
@@ -241,7 +241,12 @@ def utf16_to_utf8(obj)
     end
 
     def root
-      @root ||= @objects.deref(@objects.trailer[:Root])
+      return @root if @root
+      @root = @objects.deref(@objects.trailer[:Root])
+      unless @root.kind_of?(::Hash)
+        raise MalformedPDFError, "PDF malformed, trailer Root should be a dictionary"
+      end
+      @root
```

Comment:
To avoid the early return and instance variable dance, are you open to something like this?

```diff
diff --git a/lib/pdf/reader.rb b/lib/pdf/reader.rb
index 9a87ca8..6ce7986 100644
--- a/lib/pdf/reader.rb
+++ b/lib/pdf/reader.rb
@@ -241,12 +241,13 @@ module PDF
     end

     def root
-      return @root if @root
-      @root = @objects.deref(@objects.trailer[:Root])
-      unless @root.kind_of?(::Hash)
-        raise MalformedPDFError, "PDF malformed, trailer Root should be a dictionary"
+      @root ||= begin
+        obj = @objects.deref(@objects.trailer[:Root])
+        unless obj.kind_of?(::Hash)
+          raise MalformedPDFError, "PDF malformed, trailer Root should be a dictionary"
+        end
+        obj
       end
-      @root
```

Using `Object#tap` might clean it up even more, but it's not available all the way back to 1.9.

Reply:
👍